### PR TITLE
error handling w/o error slices and more

### PR DESCRIPTION
**Pull request by jxsl13**
poc implementation for #155 & #117

benchmarks were broken of which only one is left broken.

Some findings that I had during my implementation is that there seem to be some Setter methods defined for value receiver methods (and some more staticcheck warnings)

And you need to dig pretty deep in order to find your configuration defaults. (http.Get for example)

Overall great test coverage which helped a lot.

Edit, just saw the other PR:
#123 seems to wrap errors already. 
This PR does slightly more. 
It allows to control circular reference errors at document object contruction for the top level library.

TODO:
~~use a tool like https://pkg.go.dev/golang.org/x/exp/cmd/gorelease to see how much the api changed.~~

### CODE REVIEW

**Comment by daveshanley** [comment]
This will take some time to review! :)

**Comment by jxsl13** [comment]
Another concern that I am not yet tackling neatly is how a user can easily work with those returned errors because errors.As or errors.Is woukd detect there to be a cyclic reference error inside of that slice but you would also need a (preferrably easy)  way to check of there are also other errors inside of that multi error.

**Comment by jxsl13** [comment]
Allowed reference resolving by default

This makes the user handle errors differently.
The user, depending in their requirement, will either want or not want to resolve circular references.
In case that the user decides to forbid resolving circular references, they will handle returned errors as such and they will not need to pick out any specific error from the returned multi error.

Changing this default removes the need to actually check if one of the errors is the circular reference error.

File: index/find_component.go [hunk 1]
```diff
@@ -4,454 +4,454 @@
...
+			if err != nil {
+				e := fmt.Errorf("unable to read file bytes: %s", err)
+				return nil, nil, e
+			}
+
+		} else {
+
+			// try and read the file off the local file system, if it fails
+			// check for a baseURL and then ask our remote lookup function to go try and get it.
+			body, err = os.ReadFile(fileToRead)
```
[the remark this answers is not in the document]
Comment (jxsl13) [reply index/find_component.go:249]: false positive, filepath.Join cleans the path before passing it in os.ReadFile

**Comment by jxsl13** [comment]
dunno what problems github currently has.
On a different project of mine I did not have any such problems at all. 
Should rerun the action and ignore the security error, as  the variable comes from filepath.Join

File: index/find_component.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (daveshanley) [reply index/find_component.go:249]: Yeah, this is dumb, I don't know why it's getting upset about this one.

**Comment by daveshanley** [comment]
The tests just timeout?

**Comment by jxsl13** [comment]
the errors looked like something weird from the GitHub runner but not like the actual Go test timeout of 20 minutes was reached (which I configured).

**Comment by jxsl13** [comment]
might be that the runner or assignment to a GitHub runner timed out

**Comment by jxsl13** [comment]
first research indicates that high cpu load for an extended period of time might have caused the GitHub runner to have been canceled (crypto miner detection or whatever).

**Comment by jxsl13** [comment]
maybe it's the timeouts.

**Comment by daveshanley** [comment]
You should remove the `-race` flag. It totally crashes the runner, I am assuming these runners are provisioned with a single core and can't run race tests. Also all the race conditions have been cleaned up in #162

**Comment by jxsl13** [comment]
will merge main branch into this branch once #162 is merged into main.

**Comment by jxsl13** [comment]
@daveshanley can you trigger the tests

**Comment by jxsl13** [comment]
slightly confused as to why this happened:
```
=== RUN   TestResolver_ResolveComponents_MixedRef
    resolver_test.go:282: 
        	Error Trace:	/home/runner/work/libopenapi/libopenapi/resolver/resolver_test.go:282
        	Error:      	Not equal: 
        	            	expected: 62
        	            	actual  : 63
        	Test:       	TestResolver_ResolveComponents_MixedRef
--- FAIL: TestResolver_ResolveComponents_MixedRef (0.01s)
```

**Comment by jxsl13** [comment]
should be used as reference, errorutils might be replaces with simply errors.Join

**Pull request closed without merging**